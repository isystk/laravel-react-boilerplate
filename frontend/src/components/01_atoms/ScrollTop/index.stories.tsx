import { Meta, Story } from '@storybook/react'
import React from 'react'
import ScrollTop, { ScrollTopProps } from './index'

export default {
  title: '01_atoms/ScrollTop',
  component: ScrollTop,
} as Meta

const Template: Story = () => {
  const props: ScrollTopProps = {
    scrollY: 0,
  }
  return <ScrollTop {...props} />
}

export const Primary = Template.bind({})
Primary.storyName = 'プライマリ'
Primary.args = {}
