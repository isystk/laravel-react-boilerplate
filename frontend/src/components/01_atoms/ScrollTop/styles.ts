import { css, keyframes } from '@emotion/css'

export const scrollTop = css``

const fadeout = keyframes`
0% {
opacity: 1;
transform: translate(0, 0);
}
100% {
opacity: 0;
transform: translate(0, 50px);
}
`

export const hideButton = css`
  animation: ${fadeout} 0.5s both;
`

const fadein = keyframes`
0% {
opacity: 0;
transform: translate(0, 50px);
}
100% {
opacity: 1;
transform: translate(0, 0);
}
`

export const showButton = css`
  animation: ${fadein} 0.5s both;
`
