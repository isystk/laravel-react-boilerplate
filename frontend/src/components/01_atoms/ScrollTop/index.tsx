import React, { FC, useEffect, useState } from 'react'
import { ContainerProps } from 'types'
import { connect } from '@/components/hoc'
import * as styles from './styles'

/** ScrollTopProps Props */
export type ScrollTopProps = { scrollY }
/** Presenter Props */
export type PresenterProps = ScrollTopProps & {
  isVisible: boolean
  scrollToTop: () => void
}

/** Presenter Component */
const ScrollTopPresenter: FC<PresenterProps> = ({ isVisible, scrollToTop }) => (
  <>
    <button
      className={`${
        styles.scrollTop
      } fixed bottom-6 md:bottom-20 right-6 md:right-20 w-12 md:w-16 h-12 md:h-16 border-none bg-main rounded-full cursor-pointer transition-opacity duration-200 ease-in-out ${
        isVisible ? styles.showButton : styles.hideButton
      }`}
      onClick={scrollToTop}
    >
      <span className="text-bold text-gray-600">^</span>
    </button>
  </>
)

/** Container Component */
const ScrollTopContainer: React.FC<
  ContainerProps<ScrollTopProps, PresenterProps>
> = ({ presenter, scrollY = 500, ...props }) => {
  const [isVisible, setIsVisible] = useState(false)

  useEffect(() => {
    const toggleVisibility = () => {
      if (window.scrollY > scrollY) {
        setIsVisible(true)
      } else {
        setIsVisible(false)
      }
    }

    window.addEventListener('scroll', toggleVisibility)

    return () => window.removeEventListener('scroll', toggleVisibility)
  }, [scrollY])

  const scrollToTop = () => {
    window.scrollTo({
      top: 0,
      behavior: 'smooth',
    })
  }

  return presenter({
    isVisible,
    scrollToTop,
    ...props,
  })
}

export default connect<ScrollTopProps, PresenterProps>(
  'ScrollTop',
  ScrollTopPresenter,
  ScrollTopContainer
)
