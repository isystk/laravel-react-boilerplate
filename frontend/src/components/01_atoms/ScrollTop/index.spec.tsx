import React from 'react'
import renderer from 'react-test-renderer'
import ScrollTop, { ScrollTopProps } from './index'

describe('ScrollTop', () => {
  it('Match Snapshot', () => {
    const props: ScrollTopProps = {
      scrollY: 0,
    }
    const component = renderer.create(<ScrollTop {...props} />)
    const tree = component.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
