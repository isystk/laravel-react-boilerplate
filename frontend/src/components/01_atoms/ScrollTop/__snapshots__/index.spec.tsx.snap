// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ScrollTop Match Snapshot 1`] = `
<button
  className="css-0 fixed bottom-6 md:bottom-20 right-6 md:right-20 w-12 md:w-16 h-12 md:h-16 border-none bg-main rounded-full cursor-pointer transition-opacity duration-200 ease-in-out css-14nb29t"
  onClick={[Function]}
>
  <span
    className="text-bold text-gray-600"
  >
    ^
  </span>
</button>
`;
